packages/director/src/app.ts

```diff
@@ -95,6 +124,49 @@ app.put(
     console.log(`>> Received instance result`, { instanceId });
     await executionDriver.setInstanceResults(instanceId, result);
 
+    const instance = await executionDriver.getInstanceById(instanceId);
+    const run = await executionDriver.getRunWithSpecs(instance.runId);
+    const project = await executionDriver.getProjectById(run.meta.projectId);
+    const isRunStillRunning = run.specs.reduce(
+      (
+        wasRunning: boolean,
+        currentSpec: {
+          claimed: boolean;
+          results: any;
+        },
+        index: number
+      ) => {
+        return (
+          !currentSpec.claimed || !run.specsFull[index].results || wasRunning
+        );
+      },
+      false
+    );
+
+    reportToHook({
+      hookEvent: hookEvents.INSTANCE_FINISH,
+      reportData: {
+        run,
+        instance,
+      },
+      project,
+    }).then(() => {
+      console.log(`<< INSTANCE_FINISH hook called`, instance.instanceId);
+      // We should probably add a flag to the actual run here aswell
+      // We should also probably do a check to see if all specs passed and set a flag of success or fail
+      if (!isRunStillRunning) {
+        reportToHook({
+          hookEvent: hookEvents.RUN_FINISH,
+          reportData: {
+            run,
+            instance,
+          },
+          project,
+        });
+        console.log(`<< RUN_FINISH hook called`, run.runId);
+      }
+    });
+
```

Comment:
This new logic seems like a side effect - another argument for using some kind of emitter

Reply:
Maybe you can explain what your thinking a little more here.

Side effect is one of the newer hipster terms that tends to be used as a ... "this thing runs async code and sets state somewhere"
If thats what your meaning then... yes this runs async code. If not please correct me.

What about an emitter would make this better? 

Sorry if this comes off as abrasive. That not what im going for I just want to understand. And get on the same page.

Reply:
Yeah, sorry for dropping buzzwords. I just wanted to separate hooks from the rest of functionality in `app.ts`

> Sorry if this comes off as abrasive. 

Code reviews are a subtle art :-) You are doing great 👍🏻

Reply:
> Code reviews are a subtle art

Well said! :)